### PR TITLE
[10.x.x] Fixing material leaks in URP's renderers, features and passes

### PR DESCRIPTION
**Pull request by ellioman**
# Purpose of this PR
URP often recreates renderers, passes and renderer features, which in many cases result in Materials getting recreated as well. Unfortunately the materials were often not cleaned up resulting in sometimes massive amount of material leaking.

Quick test of this, by going in and out of playmode in the editor, resulted in:
- **URP's Template Scene:** 49 materials leaked
- **URP's Graphics Test Project:** 322 materials leaked

This PR fixes that issue by making sure URP cleans up before recreating things. Note that some of the fixes are specific to URP's test project. **With this PR no materials should leak**.

# QA
### What has been tested
- Tested in Editor (OS X 10.14.6) with URP's Template scene and URP's Graphics Test Project
- Ran the Test Runner and made sure all tests were a success
- Ran Yamato ABV + Android (GLES3 + Vulkan) + iPhone

### What needs testing
- Make sure the fix prevents materials from leaking (I've only tested in editor on OS X)
- Make sure our main test projects look and run the same (Boat attack, Viking Village, PolygonFarm)
- Make sure Post-Processing and ColorGradingLUT have not regressed
- Make sure the changes to XR have not regressed anything
- Make sure the changes to 2D have not regressed anything

# Yamato
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fmaterial-leak-fix/.yamato%252F_abv.yml%2523all_project_ci_trunk/3421657/job/pipeline

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 1]
```diff
@@ -312,7 +312,7 @@ public void Dispose()
             // Dispose all renderer features...
             for (int i = 0; i < m_RendererFeatures.Count; ++i)
             {
-                rendererFeatures[i].Dispose();
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:315]: This doesn't work properly with the Unity Object null check, should be replace with an actual `rendererFeatures[i] == null` check.

**Review by pbbastian — CHANGES_REQUESTED** [review]

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 2]
```diff
@@ -312,7 +312,7 @@ public void Dispose()
...
+                rendererFeatures[i]?.Dispose();
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:315]: This doesn't work properly with the Unity Object null check, should be replace with an actual `rendererFeatures[i] == null` check.

**Review by pbbastian — CHANGES_REQUESTED** [review]

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 2 above)
Comment (ellioman) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs]: Fixed!

**Review by pbbastian — APPROVED** [review]

**Review by thomas-zeng — APPROVED** [review]

**Review by ryanhy-unity — APPROVED** [review]
Ran through stereo tests using my Quest and all the shaders looked fine.  Doesn't seem to have any XR regressions.

**Review by simon-engelbrecht-soerensen — APPROVED** [review]
Confirmed that it prevents URP materials from leaking on Windows.
Tested with all main test projects.
Post processing works as expected and there does not appear to be any regressions.

**Review by rainsing — APPROVED** [review]

**Pull request merged**